bitcoin/src/blockdata/transaction.rs

```diff
@@ -993,6 +993,51 @@ impl Transaction {
         // `Transaction` docs for full explanation).
         self.input.is_empty()
     }
+
+    /// Returns a reference to the input at `input_index` if it exists.
+    #[inline]
+    pub fn txin(&self, input_index: usize) -> Result<&TxIn, IndexOutOfBoundsError> {
+        if input_index >= self.input.len() {
+            return Err(IndexOutOfBoundsError {
+                index: input_index,
+                length: self.input.len(),
+            });
+        }
+        Ok(&self.input[input_index])
+    }
+
+    /// Returns a reference to the output at `output_index` if it exists.
+    #[inline]
+    pub fn txout(&self, output_index: usize) -> Result<&TxOut, IndexOutOfBoundsError> {
+        if output_index >= self.output.len() {
+            return Err(IndexOutOfBoundsError {
+                index: output_index,
+                length: self.output.len(),
+            });
+        }
+        Ok(&self.output[output_index])
+    }
+}
+
+/// Error when attempting to do an out of bounds access on transaction inputs or outputs.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct IndexOutOfBoundsError {
+    /// Attempted index access.
+    pub index: usize,
+    /// Length of the vector where access was attempted.
+    pub length: usize,
+}
+
+impl fmt::Display for IndexOutOfBoundsError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "index {} is out-of-bounds for transaction input vector length {}", self.index, self.length)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for IndexOutOfBoundsError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
```

Comment:
Are you saying we should leave it out (using the default impl) or that we shouldn't implement `std::Error` at all?

I think we have a soft policy (unsure if it's written down) to never use the default impl, specifically so that with enums we'll get compilation errors when we add variants and won't forget. I guess it's reasonable to say that this rule doesn't apply to structs.

Reply:
Yeah I also thought we had a soft policy, I was thinking that being explicit at the cost of a single line makes the code more readable for newer devs, especially application developers who mightn't be as familiar with error boilerplate. For structs I'm not bothered either way, but I'd like a decision so we can be uniform.

Reply:
Leave out `source`.

> specifically so that with enums we'll get compilation errors when we add variants and won't forget. I guess it's reasonable to say that this rule doesn't apply to structs.

Exactly this.

> more readable for newer devs, especially application developers who mightn't be as familiar with error

I don't see this as worth it and it may actually confuse the developers ("why is is this reimplementing the same thing?"). Also in the past needlessly overriding defaults caused some other problem (I don't remember what it was).

I might consider it in some cases if the impl was `let &Self { field1: _, field2: _ } = self; None` so it prevents forgetting to update it if we add the source (or change to enum). This case of a clearly leaf error isn't one of them.